knowledge_storm/rm.py

```diff
@@ -396,11 +426,130 @@ def forward(self, query_or_queries: Union[str, List[str]], exclude_urls: List[st
+                )
+
+        return collected_results
+
+
+class SerperRM(dspy.Retrieve):
+    """Retrieve information from custom queries using Serper.dev.
+
+    To be compatible with STORM, the results should have the following fields:
+        - snippet: Snippets that will be used for the document
+        - title: The title of the document.
+        - url: The URL of the document. STORM use url as the unique identifier of the document, so ensure different
+            documents have different urls.
+        - description (optional): The description of the document.
+
+    """
+
+    def __init__(self, serper_search_api_key=None, query_params=None):
+        super().__init__()
+        self.usage = 0
+        self.query_params = query_params
+        self.serper_search_api_key = serper_search_api_key
+        if not self.serper_search_api_key and not os.environ.get("SERPER_API_KEY"):
+            raise RuntimeError(
+                "You must supply a serper_search_api_key param or set environment variable SERPER_API_KEY"
+            )
+
+        elif self.serper_search_api_key:
+            self.serper_search_api_key = serper_search_api_key
+
+        else:
+            self.serper_search_api_key = os.environ["SERPER_API_KEY"]
+
+        self.base_url = "https://google.serper.dev"
+
+    def serper_runner(self, query_params):
+        self.search_url = f"{self.base_url}/search"
+
+        headers = {
+            "X-API-KEY": self.serper_search_api_key,
+            "Content-Type": "application/json",
+        }
+
+        response = requests.request(
+            "POST", self.search_url, headers=headers, json=query_params
+        )
+
+        if response == None:
+            raise RuntimeError(
+                f"Error had occured while running the process {process_name}.\n Error is {response.reason}, had failed with status code {response.status_code}"
+            )
+
+        return response.json()
+
+    def get_usage_and_reset(self):
+        usage = self.usage
+        self.usage = 0
+        return {"SerperRM": usage}
+
+    def forward(self, query_or_queries: Union[str, List[str]], exclude_urls: List[str]):
+        """
+        Calls the API and searches for the query passed in.
+
+        Args:
+            query_or_queries (Union[str, List[str]]): The query or queries to search for.
+            exclude_urls (List[str]): Dummy parameter to match the interface. Does not have any effect.
+
+        Returns:
+            a list of Dicts, each dict has keys of 'description', 'snippets' (list of strings), 'title', 'url'
+        """
+        queries = (
+            [query_or_queries]
+            if isinstance(query_or_queries, str)
+            else query_or_queries
+        )
+
+        self.usage += len(queries)
+        self.results = []
+        collected_results = []
+        for query in queries:
+            if query == "Queries:":
+                continue
+            query_params = self.query_params
```

Comment:
It would be great if you could add a comment above this line with the link of Serper API in case someone is confused about what `q`, `type` are in the future.